File: rad/engine/02_catalog/model/retention.go
package model

import "time"

// RetentionOwnerKind identifies the class of long-lived consumer holding a
// durable resource pin.
type RetentionOwnerKind string

// Retention owner kinds are diagnostic; OwnerID supplies the exact identity.
const (
	RetentionOwnerPreparedPlan   RetentionOwnerKind = "prepared_plan"
	RetentionOwnerDataSnapshot   RetentionOwnerKind = "data_snapshot"
	RetentionOwnerReplica        RetentionOwnerKind = "replica"
	RetentionOwnerCDC            RetentionOwnerKind = "cdc"
	RetentionOwnerTransition     RetentionOwnerKind = "schema_transition"
	RetentionOwnerSchemaWorker   RetentionOwnerKind = "schema_worker"
	RetentionOwnerPhysicalReader RetentionOwnerKind = "physical_reader"
)

// RetentionResourceKind identifies the namespace in which an exact retained
// resource is addressed.
type RetentionResourceKind string

// Retention resource kinds intentionally keep catalog, data-position,
// transition, and physical-artifact horizons separate.
const (
	RetentionTableDefinition         RetentionResourceKind = "table_definition"
	RetentionWriteProtocolDefinition RetentionResourceKind = "write_protocol_definition"
	RetentionDataSnapshot            RetentionResourceKind = "data_snapshot"
	RetentionTransitionDeltas        RetentionResourceKind = "transition_deltas"
	RetentionPhysicalTable           RetentionResourceKind = "physical_table"
	RetentionPhysicalColumn          RetentionResourceKind = "physical_column"
	RetentionPhysicalIndex           RetentionResourceKind = "physical_index"
	RetentionTransitionDiagnostics   RetentionResourceKind = "transition_diagnostics"
)

// RetentionResource names exactly one durable resource protected by a pin.
// DataPosition remains opaque: its retention/release semantics belong to the
// KV backend, while the catalog records only the identity supplied by it.
type RetentionResource struct {
	Kind                    RetentionResourceKind `json:"kind"`
	TableID                 string                `json:"table_id,omitempty"`
	TableSchemaID           SchemaID              `json:"table_schema_id,omitempty"`
	ColumnID                string                `json:"column_id,omitempty"`
	IndexID                 string                `json:"index_id,omitempty"`
	DefinitionGeneration    uint64                `json:"definition_generation,omitempty"`
	WriteProtocolGeneration uint64                `json:"write_protocol_generation,omitempty"`
	TransitionID            string                `json:"transition_id,omitempty"`
	DataPosition            DataPosition          `json:"data_position,omitempty"`
}

// RetentionPin is a durable edge from a long-lived owner to one exact resource.
// The catalog store JSON-encodes it below /rad/catalog/retention_pin/{pin-id};
// these tags define internal Slate storage, not an API representation.
type RetentionPin struct {
	ID        string             `json:"id"`
	OwnerKind RetentionOwnerKind `json:"owner_kind"`
	OwnerID   string             `json:"owner_id"`
	Resource  RetentionResource  `json:"resource"`
	CreatedAt time.Time          `json:"created_at"`
}

// RetentionHorizon is one exact resource at the live-pin frontier. Multiple
// independent owners may retain the same resource.
type RetentionHorizon struct {
	Resource RetentionResource
	PinCount uint64
}

// RetentionHorizons is the derived resource-specific live-pin frontier. Exact
// resources are intentionally separate because no catalog generation can
// substitute for an opaque data position, transition delta stream, or physical
// range.
type RetentionHorizons struct {
	CatalogDefinitions    []RetentionHorizon
	DataSnapshots         []RetentionHorizon
	TransitionDeltas      []RetentionHorizon
	PhysicalArtifacts     []RetentionHorizon
	TransitionDiagnostics []RetentionHorizon
}

File: src/engine/02_catalog/model/retention.rs
use serde::{Deserialize, Serialize};

use super::{DataPosition, Timestamp, deserialize_null_default};
use crate::engine::catalog::identity::{
    ColumnId, DefinitionGeneration, IndexId, RetentionPinId, SchemaId, TableId, TransitionId,
    WriteProtocolGeneration,
};

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RetentionOwnerKind {
    PreparedPlan,
    DataSnapshot,
    Replica,
    Cdc,
    SchemaTransition,
    SchemaWorker,
    PhysicalReader,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum RetentionResourceKind {
    TableDefinition,
    WriteProtocolDefinition,
    DataSnapshot,
    TransitionDeltas,
    PhysicalTable,
    PhysicalColumn,
    PhysicalIndex,
    TransitionDiagnostics,
}

#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(deny_unknown_fields)]
pub struct RetentionResource {
    pub kind: RetentionResourceKind,
    #[serde(default, skip_serializing_if = "TableId::is_empty")]
    pub table_id: TableId,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub table_schema_id: Option<SchemaId>,
    #[serde(default, skip_serializing_if = "ColumnId::is_empty")]
    pub column_id: ColumnId,
    #[serde(default, skip_serializing_if = "IndexId::is_empty")]
    pub index_id: IndexId,
    #[serde(default, skip_serializing_if = "DefinitionGeneration::is_zero")]
    pub definition_generation: DefinitionGeneration,
    #[serde(default, skip_serializing_if = "WriteProtocolGeneration::is_zero")]
    pub write_protocol_generation: WriteProtocolGeneration,
    #[serde(default, skip_serializing_if = "TransitionId::is_empty")]
    pub transition_id: TransitionId,
    #[serde(default, skip_serializing_if = "DataPosition::is_empty")]
    pub data_position: DataPosition,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct RetentionPin {
    pub id: RetentionPinId,
    pub owner_kind: RetentionOwnerKind,
    pub owner_id: String,
    pub resource: RetentionResource,
    pub created_at: Timestamp,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RetentionHorizon {
    pub resource: RetentionResource,
    pub pin_count: u64,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct RetentionHorizons {
    pub catalog_definitions: Vec<RetentionHorizon>,
    pub data_snapshots: Vec<RetentionHorizon>,
    pub transition_deltas: Vec<RetentionHorizon>,
    pub physical_artifacts: Vec<RetentionHorizon>,
    pub transition_diagnostics: Vec<RetentionHorizon>,
}

// Durable records accept JSON null as an empty retained-resource list.
#[allow(dead_code)]
fn _decode_compatibility<'de, D>(deserializer: D) -> Result<Vec<RetentionResource>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    deserialize_null_default(deserializer)
}
